web/packages/teleport/src/Discover/Shared/AwsAccount/AwsAccount.tsx

```diff
@@ -96,36 +108,96 @@ export function AwsAccount() {
       nodesAccess.list &&
       nodesAccess.read; // Needed for TestConnection flow
   }
+  if (isAddingAwsApp) {
+    roleTemplate = integrationAndAppRW;
+    const appAccess = storeUser.getAppServerAccess();
+    hasAccess =
+      hasAccess &&
+      // required to upsert app server
+      appAccess.create &&
+      appAccess.edit &&
+      // required to list and read app servers
+      appAccess.list &&
+      appAccess.read;
+  }
 
-  const { attempt, run } = useAttempt(hasAccess ? 'processing' : '');
+  const { attempt, setAttempt } = useAttempt(hasAccess ? 'processing' : '');
+
+  // only used if resource we are trying to add is an AWS console.
+  const [fetchedApps, setFetchedApps] = useState<App[]>([]);
 
   const [awsIntegrations, setAwsIntegrations] = useState<Option[]>([]);
   const [selectedAwsIntegration, setSelectedAwsIntegration] =
     useState<Option>();
 
   useEffect(() => {
     if (hasAccess) {
-      fetchAwsIntegrations();
+      fetchAwsIntegrationsWithApps();
     }
   }, []);
 
-  function fetchAwsIntegrations() {
-    run(() =>
-      integrationService.fetchIntegrations().then(res => {
-        const options = res.items
-          .filter(i => i.kind === 'aws-oidc')
-          .map(i => ({
-            value: i,
-            label: i.name,
-          }));
-        setAwsIntegrations(options);
+  function makeAwsIntegrationOptions(res: IntegrationListResponse) {
+    const options = res.items
+      .filter(i => i.kind === 'aws-oidc')
+      .map(i => ({
+        value: i,
+        label: i.name,
+      }));
+    setAwsIntegrations(options);
 
-        // Auto select the only option.
-        if (options.length === 1) {
-          setSelectedAwsIntegration(options[0]);
-        }
-      })
-    );
+    // Auto select the only option.
+    if (options.length === 1) {
+      setSelectedAwsIntegration(options[0]);
+    }
+
+    return options;
+  }
+
+  async function fetchAwsIntegrationsWithApps() {
+    setAttempt({ status: 'processing' });
```

Comment:
I think we shouldn't mix setting state with data fetching in a single function, especially when the flow is complex.
I'd rather split it into two functions:
* the first one only fetches the data,
* the second one takes that data and works on it.

As of now, it is quite hard to understand where the state is exactly set and what happens when the function returns early.

Below is my attempt. I also changed `useAttempt` to `useAsync`, thanks to this we don't have to manually update the attempt's status.
```
Subject: [PATCH] Refactor data fetching in `AwsAccount`
---
Index: web/packages/teleport/src/Discover/Shared/AwsAccount/AwsAccount.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/teleport/src/Discover/Shared/AwsAccount/AwsAccount.tsx b/web/packages/teleport/src/Discover/Shared/AwsAccount/AwsAccount.tsx
--- a/web/packages/teleport/src/Discover/Shared/AwsAccount/AwsAccount.tsx	(revision def27bc646e4a1404d5bd7e586013072e46bed07)
+++ b/web/packages/teleport/src/Discover/Shared/AwsAccount/AwsAccount.tsx	(date 1715853373751)
@@ -16,7 +16,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import React, { useEffect, useState } from 'react';
+import React, { useEffect, useState, useCallback } from 'react';
 import { Link } from 'react-router-dom';
 import {
   Box,
@@ -28,19 +28,17 @@
   Flex,
 } from 'design';
 import FieldSelect from 'shared/components/FieldSelect';
-import useAttempt from 'shared/hooks/useAttemptNext';
 import { Option as BaseOption } from 'shared/components/Select';
 import Validation, { Validator } from 'shared/components/Validation';
 import { requiredField } from 'shared/components/Validation/rules';
 import TextEditor from 'shared/components/TextEditor';
-import { getErrMessage } from 'shared/utils/errorType';
+import { useAsync } from 'shared/hooks/useAsync';
 
 import { App } from 'teleport/services/apps';
 import cfg from 'teleport/config';
 import {
   Integration,
   IntegrationKind,
-  IntegrationListResponse,
   integrationService,
 } from 'teleport/services/integrations';
 import {
@@ -121,84 +119,32 @@
       appAccess.read;
   }
 
-  const { attempt, setAttempt } = useAttempt(hasAccess ? 'processing' : '');
-
-  // only used if resource we are trying to add is an AWS console.
-  const [fetchedApps, setFetchedApps] = useState<App[]>([]);
-
-  const [awsIntegrations, setAwsIntegrations] = useState<Option[]>([]);
   const [selectedAwsIntegration, setSelectedAwsIntegration] =
     useState<Option>();
 
-  useEffect(() => {
-    if (hasAccess) {
-      fetchAwsIntegrationsWithApps();
-    }
-  }, []);
+  const clusterId = storeUser.getClusterId();
 
-  function makeAwsIntegrationOptions(res: IntegrationListResponse) {
-    const options = res.items
-      .filter(i => i.kind === 'aws-oidc')
-      .map(i => ({
-        value: i,
-        label: i.name,
-      }));
-    setAwsIntegrations(options);
-
-    // Auto select the only option.
-    if (options.length === 1) {
-      setSelectedAwsIntegration(options[0]);
-    }
-
-    return options;
-  }
-
-  async function fetchAwsIntegrationsWithApps() {
-    setAttempt({ status: 'processing' });
-    try {
-      let integrationOptions = awsIntegrations;
-      // checking length, if re-attempting from failure
-      if (integrationOptions.length === 0) {
-        const integrationPage = await integrationService.fetchIntegrations();
-        integrationOptions = makeAwsIntegrationOptions(integrationPage);
-      }
-      if (!isAddingAwsApp || integrationOptions.length === 0) {
-        setAttempt({ status: 'success' });
-        // Skip fetching for apps
-        return;
-      }
-
-      // checking length, if re-attempting from failure
-      if (fetchedApps.length === 0) {
-        const resourceSvc = new ResourceService();
-        // fetch for apps that match fetched integration names.
-        // used later to determine if the integration user selected
-        // already has a application created for it.
-        const query = integrationOptions
-          .map(i => i.value.name)
-          .map(
-            integrationName =>
-              `resource.spec.integration == "${integrationName}"`
-          )
-          .join(' || ');
-
-        const { agents: resources } = await resourceSvc.fetchUnifiedResources(
-          storeUser.getClusterId(),
-          {
-            query,
-            limit: integrationOptions.length,
-            kinds: ['app'],
-            sort: { fieldName: 'name', dir: 'ASC' },
-          }
+  const [attempt, fetch] = useAsync(
+    useCallback(async () => {
+      const response = await fetchAwsIntegrationsWithApps(
+        clusterId,
+        isAddingAwsApp
+      );
+      // Auto select the only option.
+      if (response.awsIntegrations.length === 1) {
+        setSelectedAwsIntegration(
+          makeAwsIntegrationOption(response.awsIntegrations[0])
         );
-        setFetchedApps(resources.filter(r => r.kind === 'app'));
-        setAttempt({ status: 'success' });
       }
-    } catch (err) {
-      const errMsg = getErrMessage(err);
-      setAttempt({ status: 'failed', statusText: errMsg });
+      return response;
+    }, [clusterId, isAddingAwsApp])
+  );
+
+  useEffect(() => {
+    if (hasAccess && attempt.status === '') {
+      fetch();
     }
-  }
+  }, [attempt.status, fetch, hasAccess]);
 
   if (!hasAccess) {
     return (
@@ -224,7 +170,7 @@
     );
   }
 
-  if (attempt.status === 'processing') {
+  if (attempt.status === '' || attempt.status === 'processing') {
     return (
       <Box maxWidth="700px">
         <Heading />
@@ -235,12 +181,12 @@
     );
   }
 
-  if (attempt.status === 'failed') {
+  if (attempt.status === 'error') {
     return (
       <Box maxWidth="700px">
         <Heading />
         <Alert kind="danger" children={attempt.statusText} />
-        <ButtonPrimary mt={2} onClick={fetchAwsIntegrationsWithApps}>
+        <ButtonPrimary mt={2} onClick={fetch}>
           Retry
         </ButtonPrimary>
       </Box>
@@ -252,9 +198,13 @@
       return;
     }
 
-    if (isAddingAwsApp && fetchedApps.length > 0) {
+    if (
+      isAddingAwsApp &&
+      attempt.status === 'success' &&
+      attempt.data.apps.length > 0
+    ) {
       // See if an application already exists for selected integration
-      const foundApp = fetchedApps.find(
+      const foundApp = attempt.data.apps.find(
         // app => app.name === selectedAwsIntegration.value.name
         app =>
           app.integration === selectedAwsIntegration.value.name &&
@@ -281,6 +231,7 @@
     nextStep();
   }
 
+  const { awsIntegrations } = attempt.data;
   const hasAwsIntegrations = awsIntegrations.length > 0;
 
   // When a user clicks to create a new AWS integration, we
@@ -319,7 +270,7 @@
                       isSimpleValue
                       value={selectedAwsIntegration}
                       onChange={i => setSelectedAwsIntegration(i as Option)}
-                      options={awsIntegrations}
+                      options={awsIntegrations.map(makeAwsIntegrationOption)}
                     />
                   </Box>
                   <ButtonText as={Link} to={locationState} pl={0}>
@@ -360,3 +311,46 @@
     </HeaderSubtitle>
   </>
 );
+
+async function fetchAwsIntegrationsWithApps(
+  clusterId: string,
+  isAddingAwsApp: boolean
+): Promise<{
+  awsIntegrations: Integration[];
+  apps: App[];
+}> {
+  const integrationPage = await integrationService.fetchIntegrations();
+  const awsIntegrations = integrationPage.items.filter(
+    i => i.kind === 'aws-oidc'
+  );
+  if (!isAddingAwsApp || awsIntegrations.length === 0) {
+    // Skip fetching for apps
+    return { awsIntegrations, apps: [] };
+  }
+
+  const resourceSvc = new ResourceService();
+  // fetch for apps that match fetched integration names.
+  // used later to determine if the integration user selected
+  // already has an application created for it.
+  const query = awsIntegrations
+    .map(i => `resource.spec.integration == "${i.name}"`)
+    .join(' || ');
+
+  const { agents: resources } = await resourceSvc.fetchUnifiedResources(
+    clusterId,
+    {
+      query,
+      limit: awsIntegrations.length,
+      kinds: ['app'],
+      sort: { fieldName: 'name', dir: 'ASC' },
+    }
+  );
+  return { awsIntegrations, apps: resources as App[] };
+}
+
+function makeAwsIntegrationOption(integration: Integration): Option {
+  return {
+    value: integration,
+    label: integration.name,
+  };
+}
```

Reply:
thank you 🙏 , i should really switch over to `useAsync`